### PR TITLE
[3/3] Timer refactor: Timer driver mode, ETM support and erasure

### PR DESCRIPTION
**Pull request by MabezDev**
We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

This PR adds the `Mode` param to the top level timer drivers, with the various EH and inherent implementations on them. It re-adds the tests, and moves most the of testing logic to the top level driver, we now just pass in the various timer types from the different peripherals to test.

ETM for the timers has had some small changes, etm events are now created upfront before passing into the timer driver. I think we'll need a better solution for this long term, but its better than what we had and it's more consistent between the two timers now.

Closes https://github.com/esp-rs/esp-hal/issues/2321
Closes https://github.com/esp-rs/esp-hal/issues/2549
Closes https://github.com/esp-rs/esp-hal/issues/2550

### CODE REVIEW

File: esp-hal/MIGRATING-0.22.md [hunk 1]
```diff
@@ -143,4 +172,11 @@ is enabled. To retrieve the address and size of the initialized external memory,
 +let (start, size) = esp_hal::psram::psram_raw_parts(&peripherals.PSRAM, psram);
 ```
 
-The usage of `esp_alloc::psram_allocator!` remains unchanged.
+## Interrupt related reshuffle
+
+```diff
+- use esp_hal::InterruptConfigurable;
```
Comment (bugadani) [inline esp-hal/MIGRATING-0.22.md:178]: Can we land this separately?

Comment (MabezDev) [reply esp-hal/MIGRATING-0.22.md:178]: Of course, I'll pull this out of this PR.

**Comment by MabezDev** [comment]
Skip changelog for esp-wifi

File: esp-hal/MIGRATING-0.22.md [hunk 2]
```diff
@@ -140,6 +151,24 @@ Timer group timers have been type erased.
 + timg::Timer
 ```
 
+The usage of `esp_alloc::psram_allocator!` remains unchanged.
+### ETM usage has changed
+
+Timer dependant ETM events should be created _prior_ to initializing the timer with the chosen driver.
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply esp-hal/MIGRATING-0.22.md:157]: Well it's not possible to create the events once they're passed into the driver, so I'm not sure the wording matters much here :D

**Comment by bjoernQ** [comment]
I can reproduce the failing tests on H2 locally so maybe some bug sneaked in

File: esp-hal/MIGRATING-0.22.md [hunk 3]
```diff
@@ -140,6 +151,24 @@ Timer group timers have been type erased.
 + timg::Timer
 ```
 
+The usage of `esp_alloc::psram_allocator!` remains unchanged.
```
Comment (bjoernQ) [inline esp-hal/MIGRATING-0.22.md:154]: something weird happend to the file - this seems to be moved here from the PSRAM chapter and the whole `embedded-hal 0.2.* is not supported anymore` chapter disappeared

File: examples/src/bin/etm_timer.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
+//! Control LED by the boot button via ETM without involving the CPU.
+
+//! The following wiring is assumed:
+//! - LED => GPIO1
```
Comment (bjoernQ) [inline examples/src/bin/etm_timer.rs:4]: this comment is a lie 😆 it's GPIO2

File: esp-hal/src/timer/systimer.rs [hunk 5]
```diff
@@ -723,33 +728,64 @@ pub mod etm {
...
+    //! # }
+    //! ```
 
     use super::*;
 
     /// An ETM controlled SYSTIMER event
-    pub struct Event<'a> {
-        alarm: &'a mut Alarm,
+    pub struct Event {
+        id: u8,
```
Comment (bjoernQ) [inline esp-hal/src/timer/systimer.rs:772]: I wonder if s.th. like dropping the alarm could cause any harm. I guess not and holding `&'a mut Alarm` was over cautious (and annoying)

Comment (MabezDev) [reply esp-hal/src/timer/systimer.rs:772]: If the alarm is dropped and the timer is stopped, the `Event` won't occur but I think this behavior is quite obvious. The `Event` itself is essentially inert but to create it it we need temporary access to the alarm to generate the correct id.

Comment (bjoernQ) [reply esp-hal/src/timer/systimer.rs:771]: basically what I thought - I don't think we disable alarms when they get dropped but that's not in scope of this PR

Comment (MabezDev) [reply esp-hal/src/timer/systimer.rs:771]: Yeah we don't yet, but we probably will in the future

**Review by bjoernQ — APPROVED** [review]
Thanks

**Review by jessebraham — APPROVED** [review]

**Pull request merged**